### ISSUE TITLE
Multiple notification periods with different notification options

### ISSUE THREAD

**Issue by ericloyd** [NagiosEnterprises/nagioscore#95]
Okay.  This is my new favorite idea for Core:

```
Multiple notification periods with different notification options.
```

For instance:

  Normal 9-5 staff gets w/c/u/r emails delivered to 9to5@company.com
  24x7 staff gets messages gets c/r emails delivered to 24x7@company.com
  On-call staff gets c/r emails delivered outside of normal 9-5 to oncall@company.com
  No one gets emails Sundays from 00:00-03:00 for backups
  EXCEPT on-call gets emails during backups for criticals.

So, I know that this can all be done with some very complex escalations, and I know it can be done with multiple service checks checking the same thing with different options, but I think it would be cool to be able to have a multi-layer table (which is how I think of it) of who gets notified when, all on the same service, without having to switch into escalations or defining multiple services.

Discuss.  :-)

**Comment by jfrickson** [comment]
This will most likely require some changes to the notification and/or contacts and/or other objects. So it will have to wait until release 5.

**Comment by jfrickson** [comment]
Would it work to have notification options on the `contacts` and `contact_groups` options? For example:
```
define service{
  use                 local-service
  host_name           localhost
  service_description Home Partition
  check_command       check_local_disk!20%!10%!/home
  contacts            9-5staff:wcur,24x7staff:cr,on-call:cr
}
```
Or am I completely missing the point?

**Comment by ericloyd** [comment]
I like that.

**Comment by jfrickson** [comment]
@ericloyd Question:
Suppose there's a contact group named 'foo'. A contact named 'bar' is part of that group, and defaults to receiving notifications for everything. Now, given this service:
```
define service{
  use                 local-service
  host_name           localhost
  service_description Home Partition
  check_command       check_local_disk!20%!10%!/home
  contact_groups      admins,foo
  contacts            bar:c
}
```
A warning is triggered. While processing contacts, it sees bar only wants criticals and skips him. But when processing contact_groups, it sees bar wants everything and adds him to the notification list.

Is this ok, or do I need to override the default settings in this case?

**Comment by ericloyd** [comment]
What if the contact_group had an indicator that determined the behavior.  So:

`contact_groups    foo`
Would do what you wrote and send the notification anyway.

`contact_groups    foo:-`
This would send to everyone except bar, since bar is only supposed to receive c's.  IT would have no effect/meaning if bar is not listed separately and behave as the default.

`contact_groups    foo:+`
This is the explicit declaration of the default action, just for completeness.

Other "flags" could be added as needed.

**Comment by jfrickson** [comment]
```
contact_groups foo:wr-
contacts bar:cr-
```
Oops. Or would the flag be only allowed on the group?

**Comment by ericloyd** [comment]
TL;DR version: only on the group.

I was thinking that it would only be on the group.  After all, the contact being listed separately with special notification options would be to specifically override that contact's notifications.  So the simplest use scenario for this is to have someone that normally would/wouldn't get a notification that now won't/will because they're listed separately.

**Comment by jfrickson** [comment]
How about an entry in `contacts`, with or without notification flags, would override anything from `contact_groups`? That would be simpler, I think.

**Comment by ericloyd** [comment]
Sure.  Since that satisfies my original request.  So, `contact_groups` will always be notified with the member contacts' normal notifications, UNLESS the `contacts` is also specified, in which case it will add to the list of contacts notified, UNLESS there is a `:wcru" flag on the contact which then overrides anything else and specifically adds that contact (or replaces it from the group membership) with the specified notifications.

So the remaining questions for me are:  do you also do this with `contact_groups` so that more than one group can be notified differently?  do you allow `r+` and `r-` type flags so that you're adding/removing notifications?

Personally, I think the answers to those two questions should be "yes."

**Comment by jfrickson** [comment]
`:wcru` flags would be allowed on either `contacts` or `contact_groups` entries.

I would rather not have`r+` and `r-` flags. It would get pretty messy.

I'll code up the "contact overrides contact_groups" stuff and commit it so you can try it out.

**Comment by ericloyd** [comment]
That makes me happy.  :-)

**Comment by jfrickson** [comment]
Great! It's what I live for ;-)

**Comment by ericloyd** [comment]
That, and bathroom breaks.

**Comment by jfrickson** [comment]
Ok, it's in the `nv5dev` branch.
